overlord/hookstate/ctlcmd/snap_management_test.go

```diff
@@ -0,0 +1,232 @@
+	info1 := snaptest.MockSnapCurrent(c, string(snapWithCompsYaml), &snap.SideInfo{
+		Revision: snap.R(1),
+	})
+	snapstate.Set(s.st, info1.InstanceName(), &snapstate.SnapState{
+		Active: true,
+		Sequence: snapstatetest.NewSequenceFromSnapSideInfos([]*snap.SideInfo{
+			{
+				RealName: info1.SnapName(),
+				Revision: info1.Revision,
+				SnapID:   "test-snap-id",
+			},
+		}),
+		Current: info1.Revision,
+	})
+
+	s.chg = s.st.NewChange("install change", "install change")
+	task := s.st.NewTask("test-task", "my test task")
+	s.chg.AddTask(task)
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "test-hook"}
+
+	var err error
+	s.mockContext, err = hookstate.NewContext(task, task.State(), setup, s.mockHandler, "")
+	c.Assert(err, IsNil)
+}
+
+func (s *installSuite) testMngmtCommand(c *C, cmd string) {
+	s.st.Lock()
+	task := s.st.NewTask("queued", "queued task")
+	s.st.Unlock()
+
+	switch cmd {
+	case "install":
+		ctlcmd.MockSnapstateInstallComponentsFunc(func(st *state.State, comps []string) (
+			[]*state.TaskSet, error) {
+			c.Check(comps, DeepEquals, []string{"comp1", "comp2"})
+			var ts state.TaskSet
+			ts.AddTask(task)
+			return []*state.TaskSet{&ts}, nil
+		})
+	case "remove":
+		ctlcmd.MockSnapstateRemoveComponentsFunc(func(st *state.State, comps []string) (
+			[]*state.TaskSet, error) {
+			c.Check(comps, DeepEquals, []string{"comp1", "comp2"})
+			var ts state.TaskSet
+			ts.AddTask(task)
+			return []*state.TaskSet{&ts}, nil
+		})
+	}
+
+	_, _, err := ctlcmd.Run(s.mockContext, []string{cmd, "test-snap+comp1", "+comp2"}, 0)
+	c.Assert(err, IsNil)
+	s.st.Lock()
+	// one is the task added in SetUpTest, the other is the queued task
+	c.Assert(len(s.chg.Tasks()), Equals, 2)
+	s.st.Unlock()
+}
+
+func (s *installSuite) TestInstallCommand(c *C) {
+	s.testMngmtCommand(c, "install")
+}
+
+func (s *installSuite) TestRemoveCommand(c *C) {
+	s.testMngmtCommand(c, "remove")
+}
+
+func (s *installSuite) testEphemeralMngmtCommand(c *C, cmd string) {
+	s.st.Lock()
+	task := s.st.NewTask("test", "test task")
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1)}
+	s.st.Unlock()
+
+	switch cmd {
+	case "install":
+		ctlcmd.MockSnapstateInstallComponentsFunc(func(st *state.State, comps []string) (
+			[]*state.TaskSet, error) {
+			c.Check(comps, DeepEquals, []string{"comp1", "comp2"})
+			var ts state.TaskSet
+			ts.AddTask(task)
+			return []*state.TaskSet{&ts}, nil
+		})
+	case "remove":
+		ctlcmd.MockSnapstateRemoveComponentsFunc(func(st *state.State, comps []string) (
+			[]*state.TaskSet, error) {
+			c.Check(comps, DeepEquals, []string{"comp1", "comp2"})
+			var ts state.TaskSet
+			ts.AddTask(task)
+			return []*state.TaskSet{&ts}, nil
+		})
+	}
+
+	var err error
+	s.mockContext, err = hookstate.NewContext(nil, s.st, setup, s.mockHandler, "")
+	c.Assert(err, IsNil)
+
+	go func() {
+		_, _, err := ctlcmd.Run(s.mockContext, []string{cmd, "test-snap+comp1", "+comp2"}, 0)
+		c.Check(err, IsNil)
```

Comment:
shouldn't we use a channel so that the test can wait for this, otherwise we might end up theoritically running cleanups while this is still going?

Reply:
Good point, fixed now